Mathlib/Data/Fintype/Pi.lean

```diff
@@ -133,3 +135,43 @@ lemma Fin.snoc_mem_piFinset_snoc_iff {n : ℕ} {α : Fin (n + 1) → Type*}
     (Sᵢ : (i : Fin n) → Finset (α i.castSucc)) (Sₙ : Finset (α <| .last n)) :
     Fin.snoc xs x ∈ Fintype.piFinset (Fin.snoc Sᵢ Sₙ) ↔ xs ∈ Fintype.piFinset Sᵢ ∧ x ∈ Sₙ := by
   simp_rw [Fin.mem_piFinset_succ_iff', init_snoc, snoc_last]
+
+lemma Finset.map_piFinSuccAboveEquiv_filter_piFinset_succAbove {n : ℕ} (k : Fin (n + 1))
+    {α : Fin (n + 1) → Type*}
+    (p : ((i : Fin n) → α (Fin.succAbove k i)) → Prop) [DecidablePred p]
+    (S : (i : Fin (n + 1)) → Finset (α i)) :
+    ((Fintype.piFinset S).filter fun r ↦ p (fun x ↦ r <| Fin.succAbove k x)).map
+      ((Equiv.piFinSuccAboveEquiv α k).toEmbedding)
+    = S k ×ˢ (Fintype.piFinset (fun x ↦ S <| Fin.succAbove k x)).filter p := by
+  congr
+  ext ⟨x, f⟩
+  simp? [Fin.forall_iff_succAbove k] says simp only [Fintype.mem_piFinset,
+      Fin.forall_iff_succAbove k, and_imp, mem_map_equiv, Equiv.piFinSuccAboveEquiv_symm_apply,
+      mem_filter, Fin.insertNth_apply_same, Fin.insertNth_apply_succAbove, mem_product]
+  tauto
+
+lemma Finset.map_piFinSuccAboveEquiv_filter_piFinset {n : ℕ} {α : Fin (n + 1) → Type*}
+    (p : ((i : Fin n) → α i.succ) → Prop) [DecidablePred p]
+    (S : (i : Fin (n + 1)) → Finset (α i)) :
+    ((Fintype.piFinset S).filter fun r ↦ p (fun x ↦ r <| Fin.succ x)).map
+      ((Equiv.piFinSuccAboveEquiv α 0).toEmbedding)
+    = S 0 ×ˢ (Fintype.piFinset (fun x ↦ S <| Fin.succ x)).filter p :=
+  Finset.map_piFinSuccAboveEquiv_filter_piFinset_succAbove 0 p S
+
+lemma Finset.filter_piFinset_eq_map_piFinSuccAboveEquiv_symm {n : ℕ} {α : Fin (n + 1) → Type*}
+    (p : ((i : Fin n) → α i.succ) → Prop) [DecidablePred p]
+    (S : (i : Fin (n + 1)) → Finset (α i)) :
```

Comment:
I find this quite a lot more readable. Also use lowercase `s` to match the rest of the file.
```suggestion
    (S : ∀ i, Finset (α i)) :
```

Reply:
Do you want me to do this everywhere a have a dependent function? Isn't it more stylish to use `∀` when and only when the return value is a Prop?

Reply:
This new `(i : I) → ` notation is very unmathematical and verbose (I already `i` is an index, no need to spend an extra 15 characters to tell me it's a `Fin (n + 1)`).